tools/xdp/s2n-quic-xdp/src/syscall.rs

```diff
@@ -0,0 +1,342 @@
+
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        let addr = NonNull::new(addr)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "mmap returned null pointer"))?;
+
+        Ok(addr)
+    }
+}
+
+/// Unmaps a mmap region
+///
+/// # Safety
+///
+/// The following must be true:
+///
+/// * The provided `addr` MUST be a mmap'd region
+/// * The original mmap length MUST match the provided `len` parameter.
+pub unsafe fn munmap(addr: NonNull<c_void>, len: usize) -> Result<()> {
+    libc!(munmap(addr.as_ptr(), len as _))?;
+    Ok(())
+}
+
+#[inline]
+fn xdp_option<Fd: AsRawFd, T: Sized>(fd: &Fd, opt: SocketOptions, value: &mut T) -> Result<usize> {
+    let mut optlen = size_of::<T>() as libc::socklen_t;
+
+    libc!(getsockopt(
+        fd.as_raw_fd(),
+        SOL_XDP,
+        opt as _,
+        value as *mut _ as _,
+        &mut optlen,
+    ))?;
+
+    Ok(optlen as usize)
+}
+
+#[inline]
+fn set_xdp_option<Fd: AsRawFd, T: Sized>(fd: &Fd, opt: SocketOptions, value: &T) -> Result<()> {
+    let optlen = size_of::<T>() as libc::socklen_t;
+
+    libc!(setsockopt(
+        fd.as_raw_fd(),
+        SOL_XDP,
+        opt as _,
+        value as *const _ as _,
+        optlen,
+    ))?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mmap::Mmap;
+    use core::ffi::CStr;
+
+    #[test]
+    fn syscall_test() {
+        // This call requires `CAP_NET_RAW`. If the test doesn't have this set, then log and skip
+        // the next calls
+        let fd = if let Ok(fd) = open() {
+            fd
+        } else {
+            assert!(
+                std::env::var("CAP_NET_RAW_ENABLED").is_err(),
+                "expected the syscall test to be executed"
+            );
+
+            use std::io::Write;
+            let _ = writeln!(
+                std::io::stdout(),
+                "WARNING: CAP_NET_RAW capabilities missing; skipping `syscall_test`",
+            );
+            return;
+        };
+
+        // the ring sizes need to be a power of 2
+        let ring_size = 32u32;
+        let frame_size = 4096;
+        let frame_count = (ring_size * 4) as usize;
+
+        // Set up the ring sizes
+        {
+            set_fill_ring_size(&fd, ring_size).unwrap();
+            set_completion_ring_size(&fd, ring_size).unwrap();
+            set_rx_ring_size(&fd, ring_size).unwrap();
+            set_tx_ring_size(&fd, ring_size).unwrap();
+        }
+
+        // Set up the UMEM region
+        let umem_len = frame_size * frame_count;
+        let umem = Mmap::new(umem_len, 0, None).unwrap();
+        {
+            let umem_conf = UmemReg {
+                addr: umem.addr().as_ptr() as *mut _ as _,
+                chunk_size: frame_size as _,
+                flags: Default::default(),
+                headroom: 0,
+                len: umem_len as _,
+            };
+
+            set_umem(&fd, &umem_conf).unwrap();
+        }
```

Comment:
This all and the setting the ring size seems complicated. I wonder if `Mmap::new` or something equivalent should do all this for you. Is there a reason why these fn are exposed individually?


Or will you be adding a safer interface later?

Reply:
This is as low-level as it gets. The high-level abstractions are pending this PR :)